### PR TITLE
Speed up the RDF code

### PR DESCRIPTION
**Pull request by PythonFZ**
#### Summary of additions and changes

* make the RDF code faster
* applying `tf.function` is possible
* add debug message on atom pairs per second

#### How to test this pull request

run an RDF

### CODE REVIEW

File: mdsuite/calculators/radial_distribution_function.py [hunk 1]
```diff
@@ -575,14 +540,27 @@ def compute_species_values(indices, atoms_per_batch, start, d_ij):
...
+                start_ = (sum(particles_list[:tuples[0]]) - start_batch, sum(particles_list[:tuples[1]]))
+                size_ = (
+                    particles_list[tuples[0]],
+                    particles_list[tuples[1]]
+                )
+
+                names = self._get_species_names(tuples)
+
+                # select the indices that are within the boundaries of the current species / molecule
+                mask_1 = tf.math.logical_and(indices[:, 0] > start_[0], indices[:, 0] < start_[0] + size_[0])
```
Comment (SamTov) [inline mdsuite/calculators/radial_distribution_function.py:555]: This is fast, but is there a faster way without the logic? I am not sure how tf handles logic but it is typically slower than building a binary mask.

**Review by SamTov — COMMENTED** [review]
Looks great.

File: mdsuite/calculators/radial_distribution_function.py (hunk 1 above)
Comment (PythonFZ) [reply mdsuite/calculators/radial_distribution_function.py]: I don't know.
```py
mask_1 = (indices[:, 0] > start[0]) & (indices[:, 0] < stop[0])
mask_2 = (indices[:, 1] > start[1]) & (indices[:, 1] < stop[1])
```

This is just another notation that yields the same speeds I tried

File: mdsuite/calculators/calculator.py [hunk 2]
```diff
@@ -35,6 +37,10 @@
 from tqdm import tqdm
 from typing import Union, List, Any
 
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
```
Comment (SamTov) [inline mdsuite/calculators/calculator.py:41]: This always causes me issues. I am not sure we should do it.

File: mdsuite/calculators/radial_distribution_function.py [hunk 3]
```diff
@@ -33,9 +34,14 @@
 
 from mdsuite.calculators.calculator import Calculator
 from mdsuite.utils.meta_functions import split_array
+from mdsuite.utils.linalg import apply_minimum_image, get_partial_triu_indices
 
 from timeit import default_timer as timer
 
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from mdsuite import Experiment
```
Comment (SamTov) [inline mdsuite/calculators/radial_distribution_function.py:43]: Same comment

**Review by SamTov — APPROVED** [review]
My only question is about the Experiment type checking stuff. It seems to cause issue for me sometimes but if it is running then it's okay.

File: mdsuite/calculators/calculator.py (hunk 2 above)
Comment (PythonFZ) [reply mdsuite/calculators/calculator.py:41]: I understand what you mean, but it helps the type checking and with `from __future__ import annotations` this should be perfectly fine. The downside of having this import looking suspiciouse I think is less important than having the benefti of correct type checking for circular imports.

**Comment by PythonFZ** [comment]
> My only question is about the Experiment type checking stuff. It seems to cause issue for me sometimes but if it is running then it's okay.

Could you maybe try to run it on your setup?
Maybe with a 3 or more component system if you have one ready, to check that it really works.
If that is resolved I'll merge it.

**Comment by PythonFZ** [comment]
One thing we need to make sure is that the dtypes are set correctly. This is currently missing. I'll suggest adding a `self.dtype` attribute to the calculator base class

**Comment by PythonFZ** [comment]
Maybe perform a double batch on the RDF data in the minibatch loop, that way we could increase the batch size, because the data gets smaller in the loop and it could speed it up. This might be more difficult with the memory management and TQDM.

**Review by SamTov — COMMENTED** [review]
It looks good but have you tested it on > 2 species?

**Comment by PythonFZ** [comment]
> It looks good but have you tested it on > 2 species?

I have not. I don't have a system on hand to test it

**Comment by SamTov** [comment]
tested IL. All works we can push

**Pull request merged**